File: BasicsWithJava/CatchTheGame/app/src/main/java/com/sumere/catchthegame/MainActivity.java
package com.sumere.catchthegame;

import androidx.appcompat.app.AppCompatActivity;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.os.Handler;
import android.view.View;
//import android.view.animation.Animation;
//import android.view.animation.TranslateAnimation;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.Random;

public class MainActivity extends AppCompatActivity {
    private int score;
    private boolean flag;
    private TextView timerText;
    private TextView scoreText;
    private Button start;
    private ImageView[] imageArray;
    private Handler handler;
    private Runnable runnable;
    private Random random;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        score = 0;
        flag = false;
        timerText = findViewById(R.id.timerLabel);
        scoreText = findViewById(R.id.scoreLabel);
        start = findViewById(R.id.startButton);
        start.setEnabled(true);
        handler = new Handler();
        imageArray = new ImageView[9];
        imageArray[0] = findViewById(R.id.imageView);
        imageArray[1] = findViewById(R.id.imageView2);
        imageArray[2] = findViewById(R.id.imageView3);
        imageArray[3] = findViewById(R.id.imageView4);
        imageArray[4] = findViewById(R.id.imageView5);
        imageArray[5] = findViewById(R.id.imageView6);
        imageArray[6] = findViewById(R.id.imageView7);
        imageArray[7] = findViewById(R.id.imageView8);
        imageArray[8] = findViewById(R.id.imageView9);
        for(ImageView image : imageArray)
            image.setVisibility(View.INVISIBLE);
        random = new Random(0);
    }

    public void addScore(View view){
        score++;
        scoreText.setText("Score: " + score);
    }
    public void startGame(View view){
        start.setEnabled(false);
        runnable = new Runnable() {
            @Override
            public void run() {
                for(ImageView image:imageArray)
                    image.setVisibility(View.INVISIBLE);
                imageArray[random.nextInt(9)].setVisibility(View.VISIBLE);
                handler.postDelayed(this,500);
            }
        };
        handler.post(runnable);
        /*
        //Animation
        TranslateAnimation mAnimation = new TranslateAnimation(0,245,30,800);
        mAnimation.setDuration(1000);
        mAnimation.setFillAfter(true);
        mAnimation.setRepeatCount(-1);
        mAnimation.setRepeatMode(Animation.REVERSE);
        duck.setAnimation(mAnimation);
        */
        new CountDownTimer(10000,1000){
            @Override
            public void onTick(long l) {
                String temp = "Timer: " + l/1000;
                timerText.setText(temp);
            }

            @Override
            public void onFinish() {
                String temp = "Timer: 0";
                timerText.setText(temp);
                AlertDialog.Builder alert = new AlertDialog.Builder(MainActivity.this);
                alert.setTitle("Game Over");
                alert.setMessage("Score: " + score + "\nTry Again");
                alert.setPositiveButton("Yes",new DialogInterface.OnClickListener(){
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i){
                        flag = true;
                        restartGame();
                    }
                });
                alert.setNegativeButton("No", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        flag = false;
                        restartGame();
                    }
                });
                alert.show();
            }
        }.start();
    }

    public void restartGame() {
        String temp;
        score = 0;
        temp = "Score: " + score;
        scoreText.setText(temp);
        temp = "Timer: " + 100;
        timerText.setText(temp);
        if (flag) {
            start.performClick();
        } else {
            start.setEnabled(true);
            for(ImageView image:imageArray)
                image.setVisibility(View.INVISIBLE);
            handler.removeCallbacks(runnable);
        }
    }
}

File: CatchTheKenny/app/src/main/java/com/sumere/catchthekenny/MainActivity.kt
package com.sumere.catchthekenny

import android.content.DialogInterface
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.CountDownTimer
import android.os.Handler
import android.os.Looper
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.core.view.isVisible
import com.sumere.catchthekenny.databinding.ActivityMainBinding
import kotlin.random.Random

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private var random = Random(0)
    private var randomNumber = 0
    private var score = 0
    private lateinit var runnable: Runnable
    private var handler = Handler(Looper.getMainLooper())
    private var time: Long = 20000

    fun setInvisible(){
        binding.imageView1.visibility = View.INVISIBLE
        binding.imageView2.visibility = View.INVISIBLE
        binding.imageView3.visibility = View.INVISIBLE
        binding.imageView4.visibility = View.INVISIBLE
        binding.imageView5.visibility = View.INVISIBLE
        binding.imageView6.visibility = View.INVISIBLE
        binding.imageView7.visibility = View.INVISIBLE
        binding.imageView8.visibility = View.INVISIBLE
        binding.imageView9.visibility = View.INVISIBLE
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(this.layoutInflater)
        setContentView(binding.root)
        binding.timeLabelView.text = "Time: ${time/1000}"
        runnable = object: Runnable{
            override fun run() {
                randomNumber = random.nextInt() % 8 + 1

                when(randomNumber){
                    1 -> binding.imageView1.visibility = View.VISIBLE
                    2 -> binding.imageView2.visibility = View.VISIBLE
                    3 -> binding.imageView3.visibility = View.VISIBLE
                    4 -> binding.imageView4.visibility = View.VISIBLE
                    5 -> binding.imageView5.visibility = View.VISIBLE
                    6 -> binding.imageView6.visibility = View.VISIBLE
                    7 -> binding.imageView7.visibility = View.VISIBLE
                    8 -> binding.imageView8.visibility = View.VISIBLE
                    9 -> binding.imageView9.visibility = View.VISIBLE
                }
                handler.postDelayed(runnable,500)
            }
        }
    }


    fun startGame(view: View){
        binding.timeLabelView.text = "Time: 20"
        binding.scoreLabelView.text = "Score: 0"
        score = 0
        binding.startButtonView.visibility = View.INVISIBLE
        binding.startButtonView.isEnabled = false
        handler.post(runnable)
        object: CountDownTimer(time,1000){
            override fun onTick(p0: Long) {
                binding.timeLabelView.text = "Time: ${p0/1000}"
            }

            override fun onFinish() {
                binding.timeLabelView.text = "Time: 0"
                handler.removeCallbacks(runnable)
                var alert = AlertDialog.Builder(this@MainActivity)
                alert.setTitle("Game Over")
                alert.setMessage("Restart?")
                alert.setNegativeButton("No",object: DialogInterface.OnClickListener{
                    override fun onClick(p0: DialogInterface?, p1: Int) {
                        binding.timeLabelView.text = "Time: ${time/1000}"
                        binding.scoreLabelView.text = "Score: 0"
                        score = 0
                        binding.startButtonView.visibility = View.VISIBLE
                        binding.startButtonView.isEnabled = true
                        setInvisible()
                    }
                })
                alert.setPositiveButton("Yes",object: DialogInterface.OnClickListener{
                    override fun onClick(p0: DialogInterface?, p1: Int) {
                        startGame(binding.startButtonView)
                    }
                })
                alert.setCancelable(false)
                alert.show()
            }
        }.start()
    }

    fun pointScored(view: View){
        if(!binding.startButtonView.isVisible){
            score += 1
            binding.scoreLabelView.text = "Score: ${score}"
        }
    }
}